### PR TITLE
Fix IVF_PQ index creation crash if one of the cluster has no value.

### LINKED ISSUE #573: Panic when creating index
Hi folks! I've got a panic when creating a new index. Here's a Google Colab link with a minimal repro in case that's easier than the script below:

[link](https://gist.github.com/cemoody/5413c9d7628e246d53e45de07ee07aef)

```
import lance
import pyarrow as pa
import numpy as np
import pandas as pd
from loguru import logger

     

n = 1_000
pid = np.arange(0, n).astype(np.uint64)

arr = np.random.randn(n, 768).astype(np.float32)
arr = np.ravel(arr).copy()
     

!rm -rdf 0.lance
     

schema = pa.schema(
    [
        pa.field("id", pa.uint64(), False),
        pa.field("vector", pa.list_(pa.float32(), 768), True),
    ]
)

table = pa.Table.from_arrays(
    [
        pa.array(pid, type=pa.uint64()),
        pa.FixedSizeListArray.from_arrays(
            pa.array(arr, type=pa.float32()), list_size=768
        ),
    ],
    schema=schema,
)

lance.write_dataset(
    table,
    "0.lance",
)


dataset = lance.dataset("0.lance")
logger.info(f"Creating index")
dataset.create_index(
    "vector",
    index_type="IVF_PQ",
    num_partitions=64,
    num_sub_vectors=16,
)
     
```

Which gives me the following errors:

```
2023-02-13 23:46:33.892 | INFO     | __main__:<module>:25 - Creating index
---------------------------------------------------------------------------
PanicException                            Traceback (most recent call last)
<ipython-input-5-aa15cb964f53> in <module>
     24 dataset = lance.dataset("0.lance")
     25 logger.info(f"Creating index")
---> 26 dataset.create_index(
     27     "vector",
     28     index_type="IVF_PQ",

/usr/local/lib/python3.8/dist-packages/lance/dataset.py in create_index(self, column, index_type, name, **kwargs)
    282         if isinstance(column, str):
    283             column = [column]
--> 284         self._ds.create_index(column, index_type, name, kwargs)
    285 
    286 

PanicException: called `Result::unwrap()` on an `Err` value: IO("task 313 panicked")
```

### PR DESCRIPTION
**Pull request by eddyxu**
Closes #573

### CODE REVIEW

**Comment by eddyxu** [comment]
Ok, it seems to be more complicated than expect. the reason it happens is that there are empty cluster in some of the PQ subvector. right now, our PQ is fixed 8 bits centroid, which means it has to have at least 256 vectors in each partition. 

More reliability, we might want to just raise an error in such case.

**Comment by eddyxu** [comment]
An explanation of why empty cluster occurs even we have more vectors than # of centroids: https://datascience.stackexchange.com/questions/44897/k-means-clustering-what-to-do-if-a-cluster-has-0-elements#:~:text=If%20you%20get%20an%20empty,run%20from%20a%20new%20initialization.

File: rust/src/index/vector/kmeans.rs [hunk 1]
```diff
@@ -54,6 +56,11 @@ pub async fn train_kmeans(
     mut rng: impl Rng,
 ) -> Result<Float32Array> {
     let num_rows = array.len() / dimension;
+    if num_rows < k as usize {
+        return Err(crate::Error::Index(format!(
+            "KMeans: can not train {k} centroids with {num_rows} vectors"
```
Comment (changhiskhan) [inline rust/src/index/vector/kmeans.rs:61]: A more helpful message here would give the user some guidance on what `k` should be

File: rust/src/utils/kmeans.rs [hunk 2]
```diff
@@ -142,6 +142,11 @@ async fn kmean_random(
         .iter()
         .map(|v| *v as u64)
         .collect::<UInt64Array>();
+    let mut nums = std::collections::HashSet::new();
+    for i in 0..chosen.len() {
```
Comment (changhiskhan) [inline rust/src/utils/kmeans.rs:146]: does `HashSet::from_iter` work?

File: rust/src/utils/kmeans.rs [hunk 3]
```diff
@@ -159,33 +167,51 @@ struct KMeanMembership {
...
+            }))
+            .map(
+                |(cluster, (data, cluster_ids, previous_centroids))| async move {
+                    tokio::task::spawn_blocking(move || {
+                        let mut sum = Float32Array::from_iter_values(
+                            (0..dimension).map(|_| 0.0).collect::<Vec<_>>(),
+                        );
+                        let mut total = 0.0;
+                        for i in 0..cluster_ids.len() {
+                            if cluster_ids[i] == cluster {
```
Comment (changhiskhan) [inline rust/src/utils/kmeans.rs:198]: can you use the same binary_search speed up?

Comment (eddyxu) [reply rust/src/utils/kmeans.rs:198]: This is `cluster_id.filter(where id = cluster)` which is `O(n)` regardless? Btw, the cluster_ids is ordered by the vector , not sorted by value

File: rust/src/utils/kmeans.rs (hunk 2 above)
Comment (eddyxu) [reply rust/src/utils/kmeans.rs:146]: Oh this was for debugging, removed .

File: rust/src/index/vector/kmeans.rs (hunk 1 above)
Comment (eddyxu) [reply rust/src/index/vector/kmeans.rs]: updated.

**Review by changhiskhan — APPROVED** [review]
Since you're checking num_rows vs k, just change the message to say choose a k that is smaller than {num_rows}

**Pull request merged**